third_party/tcplp/bsdtcp/tcp_subr.c

```diff
@@ -0,0 +1,374 @@
+ * for TCP offloading, etc. There's no host cache in TCPlp and the host system
+ * is responsible for managing TCB memory, so much of this isn't needed. I just
+ * kept (and adpated) the few parts of the function that appeared to be needed
+ * for TCPlp.
+ */
+void
+tcp_discardcb(struct tcpcb *tp)
+{
+	tcp_cancel_timers(tp);
+
+	/* Allow the CC algorithm to clean up after itself. */
+	if (CC_ALGO(tp)->cb_destroy != NULL)
+		CC_ALGO(tp)->cb_destroy(tp->ccv);
+
+	tcp_free_sackholes(tp);
+}
+
+
+ /*
+ * Attempt to close a TCP control block, marking it as dropped, and freeing
+ * the socket if we hold the only reference.
+ */
+/*
+ * samkumar: Most of this function has to do with dropping the reference to
+ * the inpcb, freeing resources at the socket layer and marking it as
+ * disconnected, and miscellaneous cleanup. I've rewritten this to do what is
+ * needed for TCP.
+ */
+struct tcpcb *
+tcp_close(struct tcpcb *tp)
+{
+	tcp_state_change(tp, TCP6S_CLOSED); // for the print statement
+	tcp_discardcb(tp);
+	// Don't reset the TCB by calling initialize_tcb, since that overwrites the buffer contents.
+	return tp;
+}
+
+/*
+ * Create template to be used to send tcp packets on a connection.
+ * Allocates an mbuf and fills in a skeletal tcp/ip header.  The only
+ * use for this function is in keepalives, which use tcp_respond.
+ */
+/* samkumar: I changed the signature of this function. Instead of allocating
+ * the struct tcptemp using malloc, populating it, and then returning it, I
+ * have the caller allocate it. This function merely populates it now.
+ */
+void
+tcpip_maketemplate(struct tcpcb* tp, struct tcptemp* t)
+{
+	tcpip_fillheaders(tp, (void *)&t->tt_ipgen, (void *)&t->tt_t);
+}
+
+/*
+ * Fill in the IP and TCP headers for an outgoing packet, given the tcpcb.
+ * tcp_template used to store this data in mbufs, but we now recopy it out
+ * of the tcpcb each time to conserve mbufs.
+ */
+/*
+ * samkumar: This has a different signature from the original function in
+ * tcp_subr.c. In particular, IP header information is filled into an
+ * otMessageInfo rather than into a struct representing the on-wire header
+ * format. Additionally, I have changed it to always assume IPv6; I removed the
+ * code for IPv4.
+ */
+void
+tcpip_fillheaders(struct tcpcb* tp, otMessageInfo* ip_ptr, void *tcp_ptr)
+{
+	struct tcphdr *th = (struct tcphdr *)tcp_ptr;
+
+	/* Fill in the IP header */
+
+    /* samkumar: The old IPv6 code, for reference. */
+	// ip6 = (struct ip6_hdr *)ip_ptr;
+	// ip6->ip6_flow = (ip6->ip6_flow & ~IPV6_FLOWINFO_MASK) |
+	//     (inp->inp_flow & IPV6_FLOWINFO_MASK);
+	// ip6->ip6_vfc = (ip6->ip6_vfc & ~IPV6_VERSION_MASK) |
+	//     (IPV6_VERSION & IPV6_VERSION_MASK);
+	// ip6->ip6_nxt = IPPROTO_TCP;
+	// ip6->ip6_plen = htons(sizeof(struct tcphdr));
+	// ip6->ip6_src = inp->in6p_laddr;
+	// ip6->ip6_dst = inp->in6p_faddr;
+
+	memset(ip_ptr, 0x00, sizeof(otMessageInfo));
```

Comment:
@samkumar this part of code is showing problems on static analysis of code. tt_ipgen has size of 40 bytes but otMessageInfo is size of 44 bytes. was this intentional to zero out tcphdr tt_t.